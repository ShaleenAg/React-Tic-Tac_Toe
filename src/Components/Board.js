import { Square } from "./Square";
import { Button } from "@mui/material";
import React, { useEffect, useState } from "react";
import WinnerModal from "./WinnerModal";
import "./style.css";
import UndoIcon from "@mui/icons-material/Undo";
import RedoIcon from "@mui/icons-material/Redo";
import RefreshIcon from "@mui/icons-material/Refresh";
import HistoryIcon from "@mui/icons-material/History";
import HistoryModal from "./HistoryModal";

export const calculateWinner = (squares) => {
    console.log(`squares is ${squares}`);
    const lines = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [2, 4, 6],
    ];
    for (let i = 0; i < lines.length; i++) {
        const [a, b, c] = lines[i];
        if (
            squares[a] != "" &&
            squares[a] === squares[b] &&
            squares[a] === squares[c]
        ) {
            return squares[a];
        }
    }
    return null;
};
const buttonStyle = {
    margin: "5px",
};
const handleArray = (arr) => {
    arr.pop();
    return arr;
};

const Board = () => {
    const [showRewardsModal, setShowRewardsModal] = useState(false);
    const [showHistoryModal, setShowHistoryModal] = useState(false);
    const [isClicked, setIsClicked] = useState(Array(9).fill(false));
    const [playerOne, setPlayerOne] = useState(0);
    const [playerTwo, setPlayerTwo] = useState(0);
    const [squares, setSquares] = useState(Array(9).fill(""));
    const [isX, setIsX] = useState(true);
    const [stepNumber, setStepNumber] = useState(0);
    const [winner, setWinner] = useState(null);
    const [history, setHistory] = useState([
        { isClicked, squares, isX, stepNumber, winner },
    ]);
    //console.log("initial history is ", history);
    const [future, setFuture] = useState([
        { isClicked, squares, isX, stepNumber, winner },
    ]);
    //console.log(`squares is ${Array.isArray(squares)}`, squares.slice());
    const handleRewardsModal = () => setShowRewardsModal(!showRewardsModal);
    const handleHistoryModal = () => setShowHistoryModal(!showHistoryModal);
    useEffect(() => {
        //alert("Use Effect triggered!!")
        if (winner) {
            setShowRewardsModal(true);
            handleWinner(winner);
        } else if (stepNumber >= 9) {
            setShowRewardsModal(true);
        }
    }, [winner, stepNumber]);
    const onClick = (i) => {
        //console.log(`history on stepnumber: ${stepNumber} is `, history);
        //console.log("On CLick");
        // checking to see if the block is already clicked
        if (isClicked[i]) return;
        // assign isClicked[i] to true to prevent further clicks
        const newclick = isClicked.slice();
        newclick[i] = true;
        //setIsClicked(newclick);
        // if winner is already there or all blocks on the board is clicked , return;
        if (winner || stepNumber > 9) {
            return;
        }
        //console.log(squares);
        const newSquares = squares.slice();

        newSquares[i] = isX ? "X" : "O";
        // setIsX(!isX); //flipping between X and O

        // setSquares(newSquares); // assigning value to each block
        // setStepNumber(stepNumber + 1); //increasing number of moves
        // setWinner(calculateWinner(newSquares)); //calculating if winner is there after current move or null
        //alert(winner)
        //console.log(`For ${stepNumber}: ${winner}`);

        setHistory([
            ...history,
            { isClicked, squares, isX, stepNumber, winner },
        ]);

        setIsClicked(newclick);
        setIsX(!isX);
        setSquares(newSquares);
        setStepNumber(stepNumber + 1);
        setWinner(calculateWinner(newSquares));
    };

    const handleUndo = () => {
        try {
            if (history[0].isClicked == undefined) return;
            const {
                isClicked: pastClicked,
                squares: pastSquares,
                isX: pastIsX,
                stepNumber: pastStepNumber,
                winner: pastWinner,
            } = history[history.length - 1];
            //console.log("inside future", future);
            setFuture([
                ...future,
                {
                    isClicked,
                    squares,
                    isX,
                    stepNumber,
                    winner,
                },
            ]);
            setIsClicked(pastClicked);
            setIsX(pastIsX);
            setSquares(pastSquares);
            setStepNumber(pastStepNumber);
            setWinner(pastWinner);
            setHistory(handleArray(history));
        } catch (e) {
            alert("No more moves left to Undo");
            console.error(e);
        }
    };
    const handleRedo = () => {
        try {
            if (future[0].isClicked == undefined) return;
            console.log(future[2]);
            const {
                isClicked: futureClicked,
                squares: futureSquares,
                isX: futureIsX,
                stepNumber: futureStepNumber,
                winner: futureWinner,
            } = future[future.length - 1];
            console.log("After future", future);

            setHistory([
                ...history,
                {
                    isClicked,
                    squares,
                    isX,
                    stepNumber,
                    winner,
                },
            ]);
            setIsClicked(futureClicked);
            setIsX(futureIsX);
            setSquares(futureSquares);
            setStepNumber(futureStepNumber);
            setWinner(futureWinner);
            setFuture(handleArray(future));
        } catch (e) {
            console.error(e);
            alert("No more moves left to Redo");
        }
    };

    const handleWinner = (winner) => {
        //alert("inside handleWinner");
        if (winner == "X") setPlayerOne(playerOne + 1);
        else setPlayerTwo(playerTwo + 1);
    };

    const handleReset = () => {
        setSquares(Array(9).fill(""));
        setIsX(true);
        setStepNumber(0);
        setWinner(null);
        setIsClicked(Array(9).fill(false));
        setHistory([{ isClicked, squares, isX, stepNumber, winner }]);
    };
    return (
        <div className="game">
            <h1 className="game-title">Tic Tac Toe !!</h1>
            <div className="board">
                {squares.map((square, i) => {
                    return (
                        <Square
                            key={i}
                            data-key={i}
                            value={square}
                            onClick={() => onClick(i)}
                            className="squares"
                            data-testid="square"
                        />
                    );
                })}
            </div>
            <div className="buttons">
                <Button
                    sx={buttonStyle}
                    variant="contained"
                    onClick={handleUndo}
                    startIcon={<UndoIcon />}
                >
                    Undo
                </Button>
                <Button
                    sx={buttonStyle}
                    variant="contained"
                    onClick={handleRedo}
                    startIcon={<RedoIcon />}
                >
                    Redo
                </Button>
                <Button
                    sx={buttonStyle}
                    variant="contained"
                    onClick={handleReset}
                    startIcon={<RefreshIcon />}
                >
                    Refresh
                </Button>

                <Button
                    sx={buttonStyle}
                    variant="contained"
                    onClick={handleHistoryModal}
                    startIcon={<HistoryIcon />}
                >
                    History
                </Button>
            </div>
            <WinnerModal
                open={showRewardsModal}
                onClose={handleRewardsModal}
                winner={winner}
            />
            <HistoryModal
                playerOne={playerOne}
                playerTwo={playerTwo}
                open={showHistoryModal}
                onClose={handleHistoryModal}
            />
        </div>
    );
};

export default Board;
