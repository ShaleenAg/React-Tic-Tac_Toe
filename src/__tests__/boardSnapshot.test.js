import {expect , test} from '@jest/globals'

import {create} from 'react-test-renderer'
import Board from "../Components/Board"

describe("snapshot testing",()=>{
    const snap = create(<Board />).toJSON();
    test("Match initial Render", ()=>{
    expect(snap).toMatchSnapshot()
    })
})