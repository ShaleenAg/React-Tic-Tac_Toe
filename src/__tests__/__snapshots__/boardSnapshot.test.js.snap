// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing Match initial Render 1`] = `
<div
  className="game"
>
  <h1
    className="game-title"
  >
    Tic Tac Toe !!
  </h1>
  <div
    className="board"
  >
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
    <button
      className="squares"
      onClick={[Function]}
    >
      
    </button>
  </div>
  <div
    className="buttons"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-zpuu4b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="UndoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </span>
      Undo
    </button>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-zpuu4b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RedoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </span>
      Redo
    </button>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-zpuu4b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RefreshIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
          />
        </svg>
      </span>
      Refresh
    </button>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-zpuu4b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HistoryIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
          />
        </svg>
      </span>
      History
    </button>
  </div>
</div>
`;
